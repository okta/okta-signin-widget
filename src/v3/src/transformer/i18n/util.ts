/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxTransaction } from '@okta/okta-auth-js';

import {
  getI18nKey,
  getI18NParams,
} from '../../../../v2/ion/i18nUtils';
import {
  FieldElement, TokenReplacement, UISchemaElement,
} from '../../types';
import { loc } from '../../util';

export const addTranslation = ({
  element,
  name,
  i18nKey,
  params,
  defaultValue = '',
  tokenReplacement,
  noTranslate,
}: {
  element: UISchemaElement;
  name: string;
  i18nKey: string;
  params?: any;
  defaultValue?: string;
  tokenReplacement?: TokenReplacement,
  noTranslate?: boolean,
}): void => {
  // TODO: change translations to required field with default value
  // eslint-disable-next-line no-param-reassign
  element.translations = element.translations || [];
  const useDefault = !i18nKey || (noTranslate && !!defaultValue);
  element.translations.push({
    name,
    i18nKey,
    value: useDefault ? defaultValue : loc(i18nKey, 'login', params, tokenReplacement),
    noTranslate,
  });
};

export const addLabelTranslationToFieldElement = (
  transaction: IdxTransaction,
  element: FieldElement,
): void => {
  const { nextStep } = transaction;
  const { relatesTo, name: stepName } = nextStep!;
  const authenticatorKey = relatesTo?.value?.key;
  const fieldName = element.options.inputMeta.name;
  const path = authenticatorKey
    ? `${stepName}.${authenticatorKey}.${fieldName}`
    : `${stepName}.${fieldName}`;
  const i18nKey = getI18nKey(path);
  const params = getI18NParams(nextStep, authenticatorKey);
  if (i18nKey || element.label) {
    const noTranslate = element.options.inputMeta.customLabel;
    addTranslation({
      element, name: 'label', i18nKey, params, defaultValue: element.label, noTranslate,
    });
  }
};

export const addOptionalLabelTranslationToFieldElement = (
  element: FieldElement,
): void => {
  if (element.options.inputMeta.required === false) {
    addTranslation({
      element, name: 'optionalLabel', i18nKey: 'oie.form.field.optional',
    });
  }
};
