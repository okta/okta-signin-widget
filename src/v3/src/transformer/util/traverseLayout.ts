/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import {
  FormBag,
  StepperLayout,
  UISchemaElement,
  UISchemaLayout,
  UISchemaLayoutType,
} from '../../types';

type PredicateFn = (uischema: UISchemaElement) => boolean;
type CallbackFn = (uischema: UISchemaElement) => void;

type Options = {
  layout: FormBag['uischema'];
  predicate: PredicateFn;
  callback: CallbackFn;
};

export const traverseLayout = (options: Options) => {
  const fn = (
    layout: FormBag['uischema'],
    predicateFn: PredicateFn,
    callback: CallbackFn,
  ) => {
    layout.elements.forEach((element) => {
      if (predicateFn(element)) {
        callback(element);
      }

      const { type } = element;

      if (type === UISchemaLayoutType.STEPPER) {
        (element as StepperLayout).elements
          .forEach((el) => fn(el, predicateFn, callback));
        return;
      }

      if ([UISchemaLayoutType.HORIZONTAL, UISchemaLayoutType.VERTICAL]
        .includes(type as UISchemaLayoutType)) {
        fn(element as UISchemaLayout, predicateFn, callback);
      }
    });
  };

  fn(options.layout, options.predicate, options.callback);
};
