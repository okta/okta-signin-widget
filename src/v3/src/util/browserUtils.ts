/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

export const isAndroid = (): boolean => (
  // Windows Phone also contains "Android"
  /android/i.test(navigator.userAgent)
    && !/windows phone/i.test(navigator.userAgent)
);

export const isIOS = (): boolean => (
  // iOS detection from: http://stackoverflow.com/a/9039885/177710
  /iPad|iPhone|iPod/i.test(navigator.userAgent)
    // iPad on iOS 13 detection
    || (navigator.userAgent.includes('Mac') && 'ontouchend' in document)
);

export const isAndroidOrIOS = (): boolean => isAndroid() || isIOS();
