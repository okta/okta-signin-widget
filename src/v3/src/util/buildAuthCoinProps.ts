/*
 * Copyright (c) 2025-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxTransaction } from '@okta/okta-auth-js';

import {
  AUTHENTICATOR_KEY, DEVICE_ENROLLMENT_TYPE, EMAIL_AUTHENTICATOR_TERMINAL_KEYS, IDX_STEP,
} from '../constants';
import { AuthCoinProps } from '../types';
import { getAuthenticatorKey } from './getAuthenticatorKey';
import { containsOneOfMessageKeys } from './idx';

export const buildAuthCoinProps = (
  transaction?: IdxTransaction | null,
): AuthCoinProps | undefined => {
  if (!transaction) {
    return undefined;
  }

  const { nextStep, messages } = transaction;
  if (containsOneOfMessageKeys(EMAIL_AUTHENTICATOR_TERMINAL_KEYS, messages)
    || nextStep?.name === IDX_STEP.CONSENT_EMAIL_CHALLENGE) {
    return { authenticatorKey: AUTHENTICATOR_KEY.EMAIL };
  }

  if (nextStep?.name === IDX_STEP.PIV_IDP) {
    return { authenticatorKey: IDX_STEP.PIV_IDP };
  }

  if (nextStep?.name === IDX_STEP.REDIRECT_IDVERIFY && nextStep?.idp?.id) {
    return { authenticatorKey: nextStep.idp.id };
  }

  // @ts-expect-error Property 'deviceEnrollment' does not exist on type 'IdxContext' ts(2339)
  if (transaction.context?.deviceEnrollment?.value?.name === DEVICE_ENROLLMENT_TYPE.ODA) {
    return { authenticatorKey: AUTHENTICATOR_KEY.OV };
  }

  const authenticatorKey = getAuthenticatorKey(transaction);
  if (authenticatorKey) {
    return {
      authenticatorKey,
      // @ts-ignore logoUri missing from interface
      url: nextStep?.relatesTo?.value.logoUri,
    };
  }

  return undefined;
};
