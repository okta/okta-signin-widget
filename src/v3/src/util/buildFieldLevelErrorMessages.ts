/*
 * Copyright (c) 2024-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { WidgetMessage } from '../types';

type FieldLevelErrorMessages = {
  errorMessage?: string;
  errorMessageList?: string[];
};

export const buildFieldLevelErrorMessages = (
  errors: WidgetMessage[] | undefined,
) : FieldLevelErrorMessages => {
  const hasErrors = typeof errors !== 'undefined' && errors.length > 0;

  if (hasErrors) {
    if (errors.length === 1) {
      const error = errors[0];
      // Covers nested messages, e.g. password requirements not met
      // The SIW does not have messages nested more than once
      if (Array.isArray(error.message)) {
        return {
          errorMessage: error.description,
          errorMessageList: error.message
            .filter((msg) => typeof msg.message === 'string')
            .map((msg) => msg.message as string),
        };
      }
      return { errorMessage: error.message };
    }
    return {
      errorMessageList: errors
        .filter((error) => typeof error.message === 'string')
        .map((error) => error.message as string),
    };
  }

  return {};
};
