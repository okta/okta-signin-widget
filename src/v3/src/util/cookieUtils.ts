/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import Cookies from 'js-cookie';

const DEFAULT_EXPIRY_DAYS = 365;
const DEFAULT_COOKIE_PATH = '/';
const LAST_USERNAME_COOKIE_NAME = 'ln';
const USER_AUTHENTICATED_COOKIE_NAME = 'isuauth';

export const setCookieString = (
  name: string,
  value: string,
  expiryDays?: number,
  path?: string,
): void => {
  Cookies.set(
    name,
    value,
    { expires: expiryDays ?? DEFAULT_EXPIRY_DAYS, path: path ?? DEFAULT_COOKIE_PATH },
  );
};

export const getCookieString = (name: string): string => Cookies.get(name) as string;

export const getCookieUserAuthenticated = (): string => getCookieString(
  USER_AUTHENTICATED_COOKIE_NAME,
);

export const removeCookie = (name: string): void => Cookies.remove(name);

export const setUsernameCookie = (username: string): void => setCookieString(
  LAST_USERNAME_COOKIE_NAME,
  username,
);

export const setCookieUserAuthenticated = (): void => {
  Cookies.set(
    USER_AUTHENTICATED_COOKIE_NAME,
    '1',
    { expires: 1, path: DEFAULT_COOKIE_PATH },
  );
};

export const getUsernameCookie = (): string => getCookieString(LAST_USERNAME_COOKIE_NAME);

export const removeUsernameCookie = (): void => removeCookie(LAST_USERNAME_COOKIE_NAME);

export const removeUserAuthenticatedCookie = (): void => removeCookie(
  USER_AUTHENTICATED_COOKIE_NAME,
);
