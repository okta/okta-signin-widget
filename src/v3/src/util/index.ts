/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

export * from './browserUtils';
export * from './buildErrorMessageIds';
export * from './clipboard';
export * from './cookieUtils';
export * from './environmentUtils';
export * from './flattenInputs';
export * from './formatError';
export * from './formUtils';
export * from './generateRandomString';
export * from './getAuthenticatorKey';
export * from './getCurrentAuthenticator';
export * from './getElementKey';
export * from './getImmutableData';
export * from './getTranslation';
export * from './getValidationMessages';
export * from './idxUtils';
export * from './isInteractiveElement';
export * from './isPasswordRecovery';
export * from './isPollingStep';
export * from './languageUtils';
export * from './locUtil';
export * from './passwordUtils';
export * from './removeFieldLevelMessages';
export * from './resetMessagesToInputs';
export * from './settingsUtils';
export * from './shouldShowCancelLink';
export * from './toNestedObject';
export * from './webauthnUtils';
