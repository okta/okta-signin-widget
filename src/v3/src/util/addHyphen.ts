/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
*/
import { KeyboardEvent } from "src/types"

export const addHyphen = (e: KeyboardEvent<HTMLInputElement | HTMLTextAreaElement>): void => {
    const currentVal = e.currentTarget.value;
    // Add hyphen after 4th character. includes() check allows deletion of hyphen and prevents double hyphens
    if (currentVal && currentVal.length === 4 && !['Backspace', 'Delete', '-'].includes(e.key)) {
      e.currentTarget.value = currentVal.concat('-');
    } else if (currentVal && currentVal.length > 4 && currentVal.charAt(4) !== '-') {
      // If user types too fast, the above concatenation can be skipped. This retroactively adds a hyphen if it wasn't added
      e.currentTarget.value = currentVal.slice(0, 4) + '-' + currentVal.slice(4);
    }
}
