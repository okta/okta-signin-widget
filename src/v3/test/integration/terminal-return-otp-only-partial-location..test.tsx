/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { setup } from './util';

import mockResponse from '../../src/mocks/response/idp/idx/introspect/terminal-return-otp-only-partial-location.json';

describe('terminal-return-otp-only-partial-location', () => {
  it('renders form', async () => {
    const { container, findByText } = await setup({ mockResponse });
    await findByText(/Your verification code/);
    expect(container).toMatchSnapshot();
  });
});
