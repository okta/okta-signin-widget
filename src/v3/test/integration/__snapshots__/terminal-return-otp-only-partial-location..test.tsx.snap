// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`terminal-return-otp-only-partial-location renders form 1`] = `
<div>
  <div
    class="MuiScopedCssBaseline-root emotion-0"
  >
    <div
      class="MuiScopedCssBaseline-root emotion-1"
    >
      <main
        class="auth-container main-container MuiBox-root emotion-2"
        data-commit="b9bbc0140703c3fbf0e2e58920362e70"
        data-version="0.0.0"
        dir="ltr"
        id="okta-sign-in"
        lang="en"
      >
        <div
          class="MuiScopedCssBaseline-root emotion-3"
        >
          <div
            class="MuiBox-root emotion-4"
          >
            <div
              class="MuiBox-root emotion-5"
            >
              <div
                class="okta-sign-in-header auth-header authCoinSpacing MuiBox-root emotion-6"
              >
                <h1
                  class="MuiTypography-root MuiTypography-h1 emotion-7"
                />
                <div
                  aria-hidden="true"
                  class="mfa-okta-email MuiBox-root emotion-8"
                  data-se="factor-beacon"
                >
                  <svg
                    aria-labelledby="mfa-okta-email"
                    fill="none"
                    height="48"
                    role="img"
                    viewBox="0 0 48 48"
                    width="48"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <title
                      id="mfa-okta-email"
                    >
                      Email Authentication
                    </title>
                    <circle
                      class="siwFillBg"
                      cx="24"
                      cy="24"
                      fill="#F5F5F6"
                      r="24"
                    />
                    <path
                      class="siwFillSecondary"
                      d="M32 18v1h5.3L32 24.29v1.42l6-6.01v12.8a1.5 1.5 0 0 1-1.5 1.5h-17a1.5 1.5 0 0 1-1.5-1.5V31h-1v1.5a2.5 2.5 0 0 0 2.5 2.5h17a2.5 2.5 0 0 0 2.5-2.5V18h-7Z"
                      fill="#A7B5EC"
                    />
                    <path
                      class="siwFillPrimaryDark"
                      d="M9 13v14.5a2.5 2.5 0 0 0 2.5 2.5h17a2.5 2.5 0 0 0 2.5-2.5V13H9Zm20.293 1-8.81 8.81L10.794 14h18.499ZM28.5 29h-17a1.5 1.5 0 0 1-1.5-1.5V14.63l10.517 9.56L30 14.707V27.5a1.5 1.5 0 0 1-1.5 1.5Z"
                      fill="#00297A"
                    />
                  </svg>
                </div>
              </div>
              <div
                class="MuiBox-root emotion-9"
              >
                <form
                  class="o-form MuiBox-root emotion-10"
                  data-se="o-form"
                  novalidate=""
                >
                  <div
                    class="MuiBox-root emotion-11"
                  >
                    <div
                      class="MuiBox-root emotion-12"
                    >
                      <div
                        class="identifier-container MuiBox-root emotion-13"
                      >
                        <div
                          class="MuiBox-root emotion-14"
                        >
                          <span
                            class="MuiBox-root emotion-15"
                          >
                            <svg
                              class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit emotion-16"
                              fill="none"
                              focusable="false"
                              role="img"
                              viewBox="0 0 24 24"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                clip-rule="evenodd"
                                d="M15.5 6.5a3.5 3.5 0 1 1-7 0 3.5 3.5 0 0 1 7 0Zm2 0a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0ZM5 23c0-2.357.694-4.363 1.886-5.762C8.064 15.856 9.782 15 12 15c2.215 0 3.934.862 5.113 2.25C18.307 18.652 19 20.66 19 23h2c0-2.722-.807-5.216-2.363-7.046C17.067 14.107 14.785 13 12 13c-2.782 0-5.064 1.097-6.636 2.941C3.806 17.77 3 20.263 3 23h2Z"
                                fill="currentColor"
                                fill-rule="evenodd"
                              />
                              <title>
                                User
                              </title>
                            </svg>
                          </span>
                          <span
                            aria-label="test@okta.com"
                            class="identifier no-translate MuiBox-root emotion-17"
                            data-se="identifier"
                            title="test@okta.com"
                          >
                            test@okta.com
                          </span>
                        </div>
                      </div>
                    </div>
                    <div
                      class="MuiBox-root emotion-12"
                    >
                      <div
                        class="MuiBox-root emotion-19"
                      >
                        <h2
                          class="MuiTypography-root MuiTypography-h4 emotion-20"
                          data-se="o-form-head"
                          tabindex="-1"
                        >
                          Your verification code
                        </h2>
                      </div>
                    </div>
                    <div
                      class="MuiBox-root emotion-12"
                    >
                      <div
                        class="MuiBox-root emotion-19"
                      >
                        <p
                          class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph emotion-23"
                          data-se="enter-code-instr"
                        >
                          Enter this code on the sign-in page.
                        </p>
                      </div>
                    </div>
                    <div
                      class="MuiBox-root emotion-12"
                    >
                      <div
                        class="MuiBox-root emotion-25"
                      >
                        <h3
                          class="MuiTypography-root MuiTypography-h3 no-translate emotion-26"
                          data-se="otp-value"
                        >
                          123456
                        </h3>
                      </div>
                    </div>
                    <div
                      class="MuiBox-root emotion-2"
                    >
                      <div
                        class="MuiBox-root emotion-19"
                      >
                        <p
                          class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph emotion-23"
                          data-se="otp-only-request-info"
                        >
                          Request from:
                        </p>
                      </div>
                    </div>
                    <div
                      class="MuiBox-root emotion-12"
                    >
                      <div
                        class="MuiBox-root emotion-31"
                        id="browser"
                      >
                        <div
                          class="MuiBox-root emotion-32"
                          data-se="icon-browser"
                        >
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit emotion-33"
                            fill="none"
                            focusable="false"
                            role="img"
                            viewBox="0 0 24 24"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              clip-rule="evenodd"
                              d="M5 2a3 3 0 0 0-3 3v7a3 3 0 0 0 3 3h4v2H6v2h7v1.5a1.5 1.5 0 0 0 1.5 1.5h6a1.5 1.5 0 0 0 1.5-1.5v-10A1.5 1.5 0 0 0 20.5 9H18V5a3 3 0 0 0-3-3H5Zm11 7V5a1 1 0 0 0-1-1H5a1 1 0 0 0-1 1v7a1 1 0 0 0 1 1h8v-2.5A1.5 1.5 0 0 1 14.5 9H16Zm-3 6h-2v2h2v-2Zm2 5v-9h5v9h-5Z"
                              fill="currentColor"
                              fill-rule="evenodd"
                            />
                            <title>
                              Browser
                            </title>
                          </svg>
                        </div>
                        <div
                          class="MuiBox-root emotion-2"
                        >
                          <span
                            class="MuiTypography-root MuiTypography-body1 emotion-35"
                            data-se="text-browser"
                          >
                            FIREFOX on Mac OS X
                          </span>
                        </div>
                      </div>
                    </div>
                    <div
                      class="MuiBox-root emotion-12"
                    >
                      <div
                        class="MuiBox-root emotion-31"
                        id="app"
                      >
                        <div
                          class="MuiBox-root emotion-32"
                          data-se="icon-app"
                        >
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit emotion-33"
                            fill="none"
                            focusable="false"
                            role="img"
                            viewBox="0 0 24 24"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              clip-rule="evenodd"
                              d="M6.5 2a4.5 4.5 0 1 0 0 9 4.5 4.5 0 0 0 0-9ZM4 6.5a2.5 2.5 0 1 1 5 0 2.5 2.5 0 0 1-5 0ZM6.5 13a4.5 4.5 0 1 0 0 9 4.5 4.5 0 0 0 0-9ZM4 17.5a2.5 2.5 0 1 1 5 0 2.5 2.5 0 0 1-5 0Zm9-11a4.5 4.5 0 1 1 9 0 4.5 4.5 0 0 1-9 0ZM17.5 4a2.5 2.5 0 1 0 0 5 2.5 2.5 0 0 0 0-5Zm0 9a4.5 4.5 0 1 0 0 9 4.5 4.5 0 0 0 0-9ZM15 17.5a2.5 2.5 0 1 1 5 0 2.5 2.5 0 0 1-5 0Z"
                              fill="currentColor"
                              fill-rule="evenodd"
                            />
                            <title>
                              Application
                            </title>
                          </svg>
                        </div>
                        <div
                          class="MuiBox-root emotion-2"
                        >
                          <span
                            class="MuiTypography-root MuiTypography-body1 emotion-35"
                            data-se="text-app"
                          >
                            my 3rd magic link spa
                          </span>
                        </div>
                      </div>
                    </div>
                    <div
                      class="MuiBox-root emotion-12"
                    >
                      <div
                        class="MuiBox-root emotion-31"
                        id="location"
                      >
                        <div
                          class="MuiBox-root emotion-32"
                          data-se="icon-location"
                        >
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit emotion-33"
                            fill="none"
                            focusable="false"
                            role="img"
                            viewBox="0 0 24 24"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              clip-rule="evenodd"
                              d="M11.17 3.473c.36-.381.64-.473.83-.473.189 0 .47.091.83.473.363.384.736.992 1.07 1.827.581 1.454.986 3.443 1.08 5.7H9.02c.094-2.257.499-4.246 1.08-5.7.334-.835.707-1.443 1.07-1.827ZM7.019 11c.094-2.467.534-4.718 1.224-6.442.128-.32.267-.627.416-.918A9.009 9.009 0 0 0 3.055 11h3.964Zm-3.964 2a9.009 9.009 0 0 0 5.604 7.36c-.15-.29-.288-.598-.416-.918-.69-1.724-1.13-3.975-1.224-6.442H3.055Zm5.966 0h5.958c-.093 2.257-.498 4.246-1.079 5.7-.334.835-.707 1.444-1.07 1.827-.36.381-.641.473-.83.473-.189 0-.47-.091-.83-.473-.363-.384-.736-.992-1.07-1.827-.581-1.454-.986-3.443-1.08-5.7Zm7.96 0c-.094 2.467-.534 4.718-1.224 6.442-.128.32-.267.627-.416.918A9.009 9.009 0 0 0 20.945 13h-3.964Zm3.964-2a9.009 9.009 0 0 0-5.604-7.36c.15.29.288.598.416.918.69 1.724 1.13 3.975 1.224 6.442h3.964ZM23 12c0 6.075-4.925 11-11 11S1 18.075 1 12 5.925 1 12 1s11 4.925 11 11Z"
                              fill="currentColor"
                              fill-rule="evenodd"
                            />
                            <title>
                              Location
                            </title>
                          </svg>
                        </div>
                        <div
                          class="MuiBox-root emotion-2"
                        >
                          <span
                            class="MuiTypography-root MuiTypography-body1 emotion-35"
                            data-se="text-location"
                          >
                            Toronto, Canada
                          </span>
                        </div>
                      </div>
                    </div>
                    <div
                      class="MuiBox-root emotion-12"
                    >
                      <div
                        class="MuiBox-root emotion-19"
                      >
                        <p
                          class="MuiTypography-root MuiTypography-body1 MuiTypography-paragraph emotion-23"
                          data-se="otp-warning"
                        >
                          If you didn’t request this code, you can ignore this message. Your account is safe and can only be accessed with this code.
                        </p>
                      </div>
                    </div>
                  </div>
                </form>
              </div>
            </div>
          </div>
        </div>
      </main>
    </div>
  </div>
</div>
`;
